### PR TITLE
Multi ASIC HLD

### PR DESCRIPTION
**Pull request by arlakshm**

### CODE REVIEW

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+        nexthop via 10.0.107.1  dev PortChannel4009 weight 1
+
+```
+- ASIC1 learns these routes from ASIC2 as route reflection is configured on ASIC2
+```
+    root@sonic:/home/admin# ip -n asic1 route show  192.168.199.193
+    192.168.199.193 proto 186 src 172.16.132.64 metric 20
+        nexthop via 10.0.107.12  dev PortChannel4007 weight 1
+```
+  - A packet with destination ip as 192.168.199.193
```
Comment (dzhangalibaba) [inline doc/multi_asic/SONiC_multi_asic_hld.md:151]: confused here, could you review this part and make the name/ip/asicX correct?  Looks the description and the picture are not matching.

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+On a multi ASIC system, each ASIC can act as 
+ - Frontend ASIC, which connects to external devices. These ASICs have some ports as frontpanel ports and some as internal ports.
+ - Backend ASIC, which have only internal links. These ASICs only connect to other Frontend ASICs.  In a chassis system, the fabric ASIC is considered as a Backend ASIC.
+
+### 2.3.2. Control Plane
+ Within a Multi ASIC SONiC system, internal control plane can be setup in following ways
+
+#### 2.3.2.1. BGP on backend ASICs
+ In this approach:
+ - BGP instance is running on all the ASICs.
```
Comment (dev-aviz) [inline doc/multi_asic/SONiC_multi_asic_hld.md:121]: Don't this proposal consume more CPU and memory resources?

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 3]
```diff
@@ -0,0 +1,937 @@
...
+    - Configure Vlan cross-connect entries to switch packets from ASIC0 to ASIC1 and vice-versa.
+  - ASIC1 and ASIC0 have a iBGP session established
+  - The prefixes from an external Router are learnt on ASIC0  and are advertised to ASIC1 and vice-versa
+  - A packet from external Routers ingressing on ASIC0 and egressing on ASIC1 is routed twice
+  - For a device with N Frontend ASICs and M Backend ASICs, the number of BGP sessions and the number of vlans used for internal communications will be
+
+    ```
+    (( N * (N-1))/2) * M                                                                               
+    ```
+#### 2.3.2.3. Comparison of both approaches
```
Comment (dev-aviz) [inline doc/multi_asic/SONiC_multi_asic_hld.md:182]: Which proposal does this design consider to implemement? or do you provide both?

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 4]
```diff
@@ -0,0 +1,937 @@
...
+In this document, we will focus on a multi-ASIC system where ASICs can support completely independent programming/forwarding.  What this means is that traffic entering and leaving the ASIC will be in standard IP packet format, even within the router.  The ingress/egress pipeline to forward a packet is not separated into different ASICs and so there is no dependency of data-plane information shared among different ASICs in order to forward a packet from one port to another. 
+
+A linux network namespace is created for every ASIC that's controlled by the same CPU.
+This creates a separate linux network stack including routes and network devices for every ASIC.
+The interfaces for a given ASIC is linked to its corresponding namespace.
+
+In a multi-ASIC system, very commonly the ASICs are physically connected in a clos fabric topology. With sonic container dockers running as separate namespaces for each ASIC in a multi-ASIC system, we can model and configure the system as if there is a spine-leaf network topology within the box.
```
Comment (dev-aviz) [inline doc/multi_asic/SONiC_multi_asic_hld.md:98]: How to configure multi-asic ? does it have a CLI? how does it configure different ASICs

Comment (dev-aviz) [inline doc/multi_asic/SONiC_multi_asic_hld.md:98]: Multi-ASIC ?? Does it support multi vendor integration ? ex: MLNX & BRCM etc.

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 5]
```diff
@@ -0,0 +1,937 @@
...
+
+minigraph.xml is used to generate the initial configuration of a multi asic platform. There is a single minigraph.xml consisting of internal ASIC connectivity view, where each ASIC is modelled as a device.
+
+Main changes in minigraph.xml for a multi-asic fixed form platform:
+- In PngDec, a new node “ChassisInternal” is added in “DeviceLinkBase” Node. “ChassisInternal” node is used to differentiate if a link is internal link or external link.  
+- In “DeviceMetadata” sections,  DeviceProperty “SubRole” identifies whether the ASIC device is ‘Frontend’ or ‘Backend’.
+- Introduced ASIC element type for device internal chip
+- There is an internal view from the minigraph file.  It gives chipset, internal connections and internal routing logic setup for SONiC multi-asic devices.
+
+ In a Multi ASIC platform, configuration is generated per ASIC.
```
Comment (dev-aviz) [inline doc/multi_asic/SONiC_multi_asic_hld.md:308]: How does this design handle warm-reboot?

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 6]
```diff
@@ -0,0 +1,937 @@
...
+
+
+### 2.3.1. ASIC roles
+
+On a multi ASIC system, each ASIC can act as 
+ - Frontend ASIC, which connects to external devices. These ASICs have some ports as frontpanel ports and some as internal ports.
+ - Backend ASIC, which have only internal links. These ASICs only connect to other Frontend ASICs.  In a chassis system, the fabric ASIC is considered as a Backend ASIC.
+
+### 2.3.2. Control Plane
+ Within a Multi ASIC SONiC system, internal control plane can be setup in following ways
```
Comment (dev-aviz) [inline doc/multi_asic/SONiC_multi_asic_hld.md:117]: Do you have any scaling numbers ? do you see any perfromance (latency & memory foot print) issues?

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 7]
```diff
@@ -0,0 +1,937 @@
...
+Ethernet-BP4  17,18,19,20      Ethernet-BP4     Eth5-ASIC0        I
+Ethernet-BP8  21,22,23,24      Ethernet-BP8     Eth6-ASIC0        I
+Ethernet-BP12 25,26,27,28      Ethernet-BP12    Eth7-ASIC0        I 
+ ```
+
+### 2.4.3. Configuration
+
+#### 2.4.3.1. Minigraph 
+
+minigraph.xml is used to generate the initial configuration of a multi asic platform. There is a single minigraph.xml consisting of internal ASIC connectivity view, where each ASIC is modelled as a device.
```
Comment (msreddypaluru) [inline doc/multi_asic/SONiC_multi_asic_hld.md:300]: Is minigraph mandatory to configure multi-asic?

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 8]
```diff
@@ -0,0 +1,937 @@
...
+  - Route reflector is configured on the backend ASICs.
+  - All the internal neighbors are configured *next-hop-self self*
+    
+### 2.4.4. Database Service
+There will be  **N+1** database containers running in a **N**-ASIC device with a single CPU. 
+
+  -  The database container running in the linux host will be called _globalDB_ container in a multi asic platform. The database tables populated here will be system wide attributes like SYSLOG, AAA, TACACS, Mgmt interface and resources like fan, psu, thermal.
+  -  The  ***database{n}*** container will be started in each namespace. _n_  here denotes the ASIC_ID, These databases will be used and updated by applications in that namespace.
+
+Each database container will have it's own **"/var/run/redis{n}/"** directory which contains the database_config.json file, redis unix socket etc. The database_config.json will no longer be a static build time file.
```
Comment (msreddypaluru) [inline doc/multi_asic/SONiC_multi_asic_hld.md:357]: As SONiC point of view, why don't use single database container to configure multi-asic ? what is the rational behind several database containers to configure multi-asic system?

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 9]
```diff
@@ -0,0 +1,937 @@
...
+        {
+            "namespace" : "asic2",
+            "include" : "../../redis2/sonic-db/database_config.json"
+        },
+    ],
+    "VERSION" : "1.0"
+}
+
+```
+**Sample database_config.json for ASIC0**
```
Comment (msreddypaluru) [inline doc/multi_asic/SONiC_multi_asic_hld.md:389]: Does the design consider multi db support? is this template consider multi db instances ?

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 10]
```diff
@@ -0,0 +1,937 @@
...
+ * config load_minigraph : No additional namespace argument added to this command. It parses the minigraph and populates the database instances present in different namesapces.
+ * config bgp : No additional namespace argument added to bgp commands. The bgp startup/shutdown all commands are applied on the external bgp sessions ( where BGP neighbors are external routers ). The commands like bgp <startup/shutdown/remove>
+      <neighbor_ip> applies to either internal/external BGP sessions as per the user input.
+ * config interface : Added an optional argument to specify the namespace [ -n namespace ]. In Multi-ASIC devices the namespace could either be taken as a user input or if not provided will be derived based on the interface name. 
+ * config vlan : Added an optional argument to specify the namespace [ -n namespace ]. In Multi-ASIC devices namespace parameter is mandatory for (add/del) of vlan and member interface.
+ * config portchannel : Added an optional argument to specify the namespace [ -n namespace ]. In Multi-ASIC devices namesapce parameter is mandatory for (add/del) of portchannel and member interface.
```
Comment (rck-innovium) [inline doc/multi_asic/SONiC_multi_asic_hld.md:469]: Does all members of a portchannel need to be from the same ASIC? i.e. can a port-channel have member ports from ASIC0 and ASIC1?  
Please document it.

File: doc/multi_asic/SONiC_multi_asic_hld.md [hunk 11]
```diff
@@ -0,0 +1,937 @@
...
+      /etc/sonic/config_db.json, /etc/sonic/config_db0.json, /etc/sonic/config_db1.json etc as there are number of asic/namespace needed in the platform.
+ * config load_minigraph : No additional namespace argument added to this command. It parses the minigraph and populates the database instances present in different namesapces.
+ * config bgp : No additional namespace argument added to bgp commands. The bgp startup/shutdown all commands are applied on the external bgp sessions ( where BGP neighbors are external routers ). The commands like bgp <startup/shutdown/remove>
+      <neighbor_ip> applies to either internal/external BGP sessions as per the user input.
+ * config interface : Added an optional argument to specify the namespace [ -n namespace ]. In Multi-ASIC devices the namespace could either be taken as a user input or if not provided will be derived based on the interface name. 
+ * config vlan : Added an optional argument to specify the namespace [ -n namespace ]. In Multi-ASIC devices namespace parameter is mandatory for (add/del) of vlan and member interface.
```
Comment (rck-innovium) [inline doc/multi_asic/SONiC_multi_asic_hld.md:468]: Can the VLAN member ports of a given VLAN be from different ASICs? 

If yes, then, say we need to do an ARP request on the VLAN interface-- how will this be achieved?

**Review by rlhui — APPROVED** [review]

**Pull request merged**